src/handlers/assign.rs

```diff
@@ -666,14 +666,112 @@ fn find_reviewer_from_names(
     // These are all ideas for improving the selection here. However, I'm not
     // sure they are really worth the effort.
+    use rand::prelude::IteratorRandom;
     Ok(candidates
         .into_iter()
         .choose(&mut rand::thread_rng())
         .expect("candidate_reviewers_from_names always returns at least one entry")
         .to_string())
 }
 
-/// Returns a list of candidate usernames to choose as a reviewer.
+async fn new_find_reviewer_from_names(
+    db_client: &DbClient,
+    candidates: &HashSet<&str>,
+) -> Result<ReviewCapacityUser, FindReviewerError> {
+    let candidates_vec = candidates
+        .iter()
+        .map(|x| x.to_string())
+        .collect::<Vec<String>>();
+    let assignee = if candidates_vec.len() > 1 {
+        // Select the best candidate from the pool
+        match get_review_candidate_by_capacity(&db_client, candidates_vec.clone()).await {
+            Ok(reviewers) => reviewers,
+            Err(_) => {
+                return Err(FindReviewerError::NoReviewer {
+                    initial: candidates_vec,
+                });
+            }
+        }
+    } else {
+        // Get the prefs of the only candidate identified
+        match get_review_candidates_by_username(&db_client, candidates_vec).await {
+            Ok(mut reviewers) => {
+                if reviewers.is_empty() {
+                    return Err(FindReviewerError::NoReviewer { initial: vec![] });
+                }
+                reviewers
+                    .pop()
+                    .expect("Something wrong happened while getting the reviewer")
+            }
+            Err(_) => {
+                return Err(FindReviewerError::NoReviewer { initial: vec![] });
+            }
+        }
+    };
+    Ok(assignee)
+}
+
+/// Finds a reviewer to assign to a PR.
+/// Accounts for reviewer's capacity preferences.
+/// If just one candidate is available (or a specific reviewer is invoked), return that one.
+///
+/// The `names` is a list of candidate reviewers `r?`, such as `compiler` or
+/// `@octocat`, or names from the owners map. It can contain GitHub usernames,
+/// auto-assign groups, or rust-lang team names. It must have at least one
+/// entry.
+async fn find_reviewer_from_names(
+    db_client: &DbClient,
+    teams: &Teams,
+    config: &AssignConfig,
+    issue: &Issue,
+    names: &[String],
+) -> Result<ReviewCapacityUser, FindReviewerError> {
+    let candidates = candidate_reviewers_from_names(teams, config, issue, names)?;
+    let assignee: Result<ReviewCapacityUser, FindReviewerError>;
+    if use_new_pr_assignment(teams, &candidates) {
+        log::debug!("Using NEW pull request assignment workflow");
+        assignee = new_find_reviewer_from_names(db_client, &candidates).await;
+    } else {
+        log::debug!("Using OLD pull request assignment workflow");
+        let username = old_find_reviewer_from_names(candidates);
+        assignee = Ok(ReviewCapacityUser::phony(username.unwrap()));
+    }
```

Comment:
The new pull request assignment can be enabled or disengaged using an env var (`USE_NEW_PR_ASSIGNMENT`) when deploying the Triagebot. This conditional allows using the current workflow or the new one.